File: exercise/tries/word-dictionary.js

class Node {
  constructor(d){
    this.d = d;
    this.map = {};
    this.terminal = false;
  }
}

class Trie {
  constructor(){
    this.root = new Node('\n')
  }
  add(words){
    let cursor = this.root;
    for(const c of words){
      if(cursor.map[c] === undefined){
        cursor.map[c] = new Node(c)
      }
      cursor = cursor.map[c];
    }
    cursor.terminal = true;
  }
  dfs(pivot, word, node){
    if(pivot >= word.length) {
      if(node.terminal) {
        return true;
      } else {
        return false;
      }
    }

    const c = word.charAt(pivot);
    // console.log(`pivot ${pivot} | c ${c}`)
    let result = false;
    if(c === '.'){
      for(const k of Object.keys(node.map)){
        const v = node.map[k];
        if(this.dfs(pivot + 1, word, v)){
          result = true;
        }
      }
    } else if(node.map[c] === undefined) {
      return false;
      } else if(this.dfs(pivot + 1, word, node.map[c])){
      result = true;
    }
    return result;
  }
  search(key){
    return this.dfs(0, key, this.root);
  }
}

let t;
var WordDictionary = function() {
  t = new Trie();
};

/** 
 * @param {string} word
 * @return {void}
 */
WordDictionary.prototype.addWord = function(word) {
  t.add(word)
};

/** 
 * @param {string} word
 * @return {boolean}
 */
WordDictionary.prototype.search = function(word) {
    return t.search(word)
};

/** 
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */

// t.add('justine')

// console.log(t.search('a'))

File: exercise/trie/word-dictionary.cpp
#include <iostream>
#include <map>
using namespace std;

class Node {
  public:
    char d;
    map<char, Node*> map;
    bool terminal;

    Node(char _d){
      d = _d;
      terminal = false;
    }
};

class WordDictionary {
public:
    Node* root;
    WordDictionary() {
      root = new Node('\0');
    }
    
    void addWord(string word) {
      Node* cursor = root;
      for(char w : word){
        if(cursor->map.count(w) == 0) {
          cursor->map[w] = new Node(w);
        }
        cursor = cursor->map[w];
      }
      cursor->terminal = true;
    }

    bool dfs(int pivot, string word, Node *node){
      if(pivot >= word.length()) {
        if(node->terminal) {
          return true;
        }
        return false; 
      }
      char c = word.at(pivot);
      bool result = false;
      if(c == '.') {
        for(auto& pair : node->map) {
          if(dfs(pivot + 1, word, pair.second)) {
            result = true;
          }
        }
      } else if(node->map.count(c) == 0){
        return false;
      } else if(dfs(pivot + 1, word, node->map[c])){
        result = true;
      }
      return result;
    }

    bool search(string word) {
      return dfs(0, word, root);
    }
};

int main(){
  WordDictionary *dict = new WordDictionary();
  dict->addWord("bad");
  dict->addWord("dad");
  dict->addWord("mad");

  cout << "Result : " << dict->search("") << "\n";
  return 0;
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */